src/on_chain/on_chain_surplus.py

```diff
@@ -0,0 +1,366 @@
+        gets response from AWS bucket and makes call to Quasimodo for required orders.
+        """
+        try:
+            encoded_transaction = self.web_3.eth.get_transaction(settlement_hash)
+            solver_address = encoded_transaction["from"]
+            print(solver_address)
+            decoded_settlement = DecodedSettlement.new(
+                self.contract_instance, encoded_transaction.input
+            )
+        except ValueError as except_err:
+            self.logger.error(
+                "Unhandled exception, possibly can't decode: %s", str(except_err)
+            )
+            return None
+        (winning_orders, bucket_response) = self.get_order_data_by_hash(settlement_hash)
+        if bucket_response is None:
+            return None
+        self.solve_orders_in_settlement(bucket_response, winning_orders, settlement_hash, decoded_settlement)
+        return "200"
+
+
+    def get_solver_response(order_id, bucket_response):
+        for key, order in bucket_response["orders"].items():
+            if order["id"] == order_id:
+                bucket_response["orders"] = {key: order}
+                break
+
+        # convert back to JSON for sending to Quasimodo
+        # bucket_response_json = json.dumps(bucket_response)
+
+        # # space here to post and receive instance JSON from Quasimodo
+        # solution_json = requests.post(bucket_response_json)
+
+        # assuming jsonObject is called solution_json
+        # solution = solutions_json.json() to convert to dict
+
+        # 'solution' is what we use here as a python object in instance_file.py
+        # return quasimodo solved solution
+        # return solution, order
+
+
+    def solve_orders_in_settlement(self, bucket_response, winning_orders, settlement_hash, decoded_settlement):
+        # there can be multiple orders/trades in a single settlement
+        Bucket_Response = deepcopy(bucket_response)
+        for trade in decoded_settlement.trades:
+            del bucket_response
+            bucket_response = deepcopy(Bucket_Response)
+            sell_token_index = trade["sellTokenIndex"]
+            buy_token_index = trade["buyTokenIndex"]
+            sell_token_clearing_price = decoded_settlement.clearing_prices[
+                sell_token_index
+            ]
+            buy_token_clearing_price = decoded_settlement.clearing_prices[
+                buy_token_index
+            ]
+            order_type = str(
+                "{0:08b}".format(trade["flags"])
+            )  # convert flags value to binary to extract L.S.B (Least Sigificant Byte)
+            winning_surplus = self.get_surplus(
+                trade,
+                sell_token_clearing_price,
+                buy_token_clearing_price,
+                order_type[-1],
+            )
+            order_id = winning_orders[decoded_settlement.trades.index(trade)]["id"]
+            solver_solution, order = self.get_solver_response(order_id, bucket_response)
```

Comment:
Is this function mutating `bucket_response`? That might be a bit surprising. In this case I would move the `deepcopy` stuff into that function.

Reply:
It is, because the orders field is updated to contain a single order, then the bucket_response is sent to quasimodo to get solutions. 
When we move to the next order, we need to find the single order that must be filtered out based on order ID. But bucket_response is only a single order, which is why a copy must be made to restore to the original response for each order. Simple assignment (=) points to the same memory location, and therefore I've created a deepcopy.